File: firmware/esp32_sensor_node/AlertManager.cpp
#include "AlertManager.h"
#include "AWSIoTManager.h"
#include "FirebaseManager.h"
#include <ArduinoJson.h>

AlertManager::AlertManager(const char* thingName, AWSIoTManager* aws, FirebaseManager* fb)
    : _thingName(thingName), _aws(aws), _fb(fb), _lastAlertMs(0)
{
    _lastAlertType[0] = '\0';
}

bool AlertManager::sendAnomaly(const char* userId, float anomalyScore,
                                const char* alertType, const char* detail) {
    uint32_t now = millis();

    // Suppress duplicate alert types within cooldown window
    if (strcmp(_lastAlertType, alertType) == 0 &&
        (now - _lastAlertMs) < ALERT_COOLDOWN_MS) {
        Serial.printf("[ALERT] Cooldown active for '%s' — suppressed\n", alertType);
        return false;
    }

    AlertPayload p;
    strlcpy(p.thingName,  _thingName,  sizeof(p.thingName));
    strlcpy(p.userId,     userId,      sizeof(p.userId));
    strlcpy(p.alertType,  alertType,   sizeof(p.alertType));
    strlcpy(p.detail,     detail ? detail : "", sizeof(p.detail));
    p.anomalyScore = anomalyScore;
    p.ts           = now;

    String json = _format(p);

    bool dispatched = false;

    // Publish to AWS IoT → triggers Lambda → Bedrock Agent → user notification
    if (_aws) {
        dispatched = _aws->publishAlertJson(json);
    }

    // Log to Firebase for history and dashboard visibility
    if (_fb) {
        _fb->pushBiometricAlert(userId, alertType, anomalyScore);
    }

    if (dispatched || _fb) {
        _lastAlertMs = now;
        strlcpy(_lastAlertType, alertType, sizeof(_lastAlertType));
        Serial.printf("[ALERT] Sent: type=%s user=%s score=%.3f\n",
                      alertType, userId, anomalyScore);
    }

    return dispatched;
}

void AlertManager::onAgentAck(const char* userId, const char* alertType) {
    Serial.printf("[ALERT] Agent ACK: type=%s user=%s — notification delivered\n",
                  alertType, userId);
    // Extend cooldown to prevent re-alerting after confirmed delivery
    _lastAlertMs = millis();
}

String AlertManager::_format(const AlertPayload& p) const {
    StaticJsonDocument<384> doc;
    doc["deviceId"]    = p.thingName;
    doc["userId"]      = p.userId;
    doc["alertType"]   = p.alertType;
    doc["anomalyScore"] = p.anomalyScore;
    doc["detail"]      = p.detail;
    doc["ts"]          = p.ts;

    String out;
    serializeJson(doc, out);
    return out;
}

File: firmware/rpi_sensor_node/alert_manager.py
import json
import time

_ALERT_COOLDOWN_MS = 30_000.0


class AlertManager:
    """
    Alert dispatcher — mirrors AlertManager.h/.cpp.
    Suppresses duplicate alert types within the cooldown window.
    Publishes to AWS IoT (→ Lambda → Bedrock Agent) and logs to Firebase.
    """

    def __init__(self, thing_name: str, aws_iot, firebase):
        self._thing_name      = thing_name
        self._aws             = aws_iot
        self._fb              = firebase
        self._last_alert_ms   = 0.0
        self._last_alert_type = ""

    def send_anomaly(self, user_id: str, anomaly_score: float,
                     alert_type: str, detail: str = "") -> bool:
        now_ms = time.time() * 1000

        if (self._last_alert_type == alert_type and
                (now_ms - self._last_alert_ms) < _ALERT_COOLDOWN_MS):
            print(f"[ALERT] Cooldown active for '{alert_type}' — suppressed")
            return False

        payload = json.dumps({
            "deviceId":    self._thing_name,
            "userId":      user_id,
            "alertType":   alert_type,
            "anomalyScore": anomaly_score,
            "detail":      detail,
            "ts":          int(now_ms),
        })

        dispatched = False
        if self._aws:
            dispatched = self._aws.publish_alert_json(payload)
        if self._fb:
            self._fb.push_biometric_alert(user_id, alert_type, anomaly_score)

        if dispatched or self._fb:
            self._last_alert_ms   = now_ms
            self._last_alert_type = alert_type
            print(f"[ALERT] Sent: type={alert_type} user={user_id} score={anomaly_score:.3f}")

        return dispatched

    def on_agent_ack(self, user_id: str, alert_type: str):
        print(f"[ALERT] Agent ACK: type={alert_type} user={user_id} — notification delivered")
        # Extend cooldown so the same alert isn't re-sent immediately after ACK
        self._last_alert_ms = time.time() * 1000
